**EXERCISE-2**

#Given Optimization problem:-

\begin{align}
\max \ 35 x_1 + 57 x_2 + 48 x_3 + 20 x_4 + 15 x_5, \\
{\rm{s.t.}} \; x_1 + x_2 + x_3 & \leq 75, \\
x_3 + x_4 + x_5 & \leq 100, \\
x_1 + x_2 & \geq 10, \\ 
30000 x_1 + 95000 x_2 + 10000 x_3 + 5000 x_4 + 4000 x_5 & \leq 1500000, \\
100000 x_1 + 500000 x_2 + 200000 x_3 + 45000 x_4 + 25000 x_5 & \geq 900000, \\
x_i & \geq 0, \quad i \in \{1,2,3,4,5\}. \\
\end{align}

Here $x_1,x_2,x_3,x_4$ and $x_5$ denote the number of different type of media slots to be used for advertisement, where $x_1$ corresponds to website, $x_2$ corresponds
to TV, $x_3$ corresponds to daily newspaper, $x_4$ corresponds to weekly magazine and $x_5$ corresponds
to monthly magazine.

#Que.1

given that:
$z_1=x_1+x_2\\ z_2=x_3+x_4+x_5$

the new constraint to be added in the problem is:$|z_1-z_2|\leq18$

In terms of the decision variables the constraint will be:
$$|x_1+x_2-x_3-x_4-x_5|\leq18$$

after adding this constraints the problem will be:
\begin{align}
\max \ 35 x_1 + 57 x_2 + 48 x_3 + 20 x_4 + 15 x_5, \\
{\rm{s.t.}} \; x_1 + x_2 + x_3 & \leq 75, \\
x_3 + x_4 + x_5 & \leq 100, \\
x_1 + x_2 & \geq 10, \\ 
30000 x_1 + 95000 x_2 + 10000 x_3 + 5000 x_4 + 4000 x_5 & \leq 1500000, \\
100000 x_1 + 500000 x_2 + 200000 x_3 + 45000 x_4 + 25000 x_5 & \geq 900000,\\|x_1+x_2-x_3-x_4-x_5|\leq18 \\
x_i & \geq 0, \quad i \in \{1,2,3,4,5\}. \\
\end{align}


Above problem is not a linear program because the constraints consist modulus function which is a piecewise function hence the problem is not linear.

#Que.2

#Approach 1:
let $|x_1+x_2-x_3-x_4-x_5|=u$ and it is clear that $u\geq0$

and the following inequalities are satisfied:

$$u \geq x_1+x_2-x_3-x_4-x_5 \\
u \geq -x_1-x_2+x_3+x_4+x_5$$ 

now after this substitution our model will be:
\begin{align}
\max \ 35 x_1 + 57 x_2 + 48 x_3 + 20 x_4 + 15 x_5, \\
{\rm{s.t.}} \; x_1 + x_2 + x_3 & \leq 75, \\
x_3 + x_4 + x_5 & \leq 100, \\
x_1 + x_2 & \geq 10, \\ 
30000 x_1 + 95000 x_2 + 10000 x_3 + 5000 x_4 + 4000 x_5 & \leq 1500000, \\
100000 x_1 + 500000 x_2 + 200000 x_3 + 45000 x_4 + 25000 x_5 & \geq 900000, \\u-x_1-x_2+x_3+x_4+x_5 \geq 0 \\
u+x_1+x_2-x_3-x_4-x_5 \geq 0,\\u\geq18\\
x_i & \geq 0, \quad i \in \{1,2,3,4,5\}. \\
\end{align}


#Approach-2

let $x_1+x_2-x_3-x_4-x_5=a-b$

then $|x_1+x_2-x_3-x_4-x_5|=|a-b|=a+b$  ;  where $a\geq0\  and\  b\geq0$
after applying this approach our model will be:
\begin{align}
\max \ 35 x_1 + 57 x_2 + 48 x_3 + 20 x_4 + 15 x_5, \\
{\rm{s.t.}} \; x_1 + x_2 + x_3 & \leq 75, \\
x_3 + x_4 + x_5 & \leq 100, \\
x_1 + x_2 & \geq 10, \\ 
30000 x_1 + 95000 x_2 + 10000 x_3 + 5000 x_4 + 4000 x_5 & \leq 1500000, \\
100000 x_1 + 500000 x_2 + 200000 x_3 + 45000 x_4 + 25000 x_5 & \geq 900000, \\
x_1+x_2-x_3-x_4-x_5-a+b=0,\\
a+b \leq 18,\\
a\geq0,\\
b\geq0,\\
x_i & \geq 0, \quad i \in \{1,2,3,4,5\}. \\
\end{align}



#Que.3

practice code of Lab4:

In [1]:
!pip install -q pyomo

     |████████████████████████████████| 9.1 MB 5.1 MB/s 
     |████████████████████████████████| 49 kB 5.0 MB/s 


In [2]:
from pyomo.environ import *

In [3]:
import pandas as pd

In [4]:
import numpy as np

In [5]:
data_csvfile = pd.read_csv('lab6_ex2.csv')

In [6]:
print (type(data_csvfile))

<class 'pandas.core.frame.DataFrame'>


In [7]:
data_csvfile.columns

Index(['Medium', 'Num Customers', 'Cost per advt', 'Max slots per month',
       'Reach Quality'],
      dtype='object')

In [8]:
data_csvfile['Medium']

0           Websites
1                 TV
2     DailyNewspaper
3     WeeklyMagazine
4    MonthlyMagazine
Name: Medium, dtype: object

In [9]:
data_csvfile['Num Customers']

0    100000
1    500000
2    200000
3     45000
4     25000
Name: Num Customers, dtype: int64

In [10]:
data_csvfile['Cost per advt']

0    30000
1    95000
2    10000
3     5000
4     4000
Name: Cost per advt, dtype: int64

In [11]:
data_csvfile['Max slots per month']

0    20
1    25
2    30
3     4
4     1
Name: Max slots per month, dtype: int64

In [12]:
data_csvfile['Reach Quality']

0    35
1    57
2    48
3    20
4    15
Name: Reach Quality, dtype: int64

In [13]:
data_csvfile.index

RangeIndex(start=0, stop=5, step=1)

In [14]:
len(data_csvfile.index)

5

In [15]:
model_lab4_ex2 = ConcreteModel()

In [16]:
N = len(data_csvfile.index)

In [17]:
M=5

In [18]:
col_indices = range(N)

In [19]:
model_lab4_ex2.x = Var(col_indices)

In [20]:
model_lab4_ex2.objective = Objective(expr=summation(data_csvfile['Reach Quality'], model_lab4_ex2.x),sense=maximize)

In [21]:
model_lab4_ex2.constraints = ConstraintList()

In [22]:
model_lab4_ex2.constraints.add(expr= model_lab4_ex2.x[0] + model_lab4_ex2.x[1] + model_lab4_ex2.x[2]<= 75  )
model_lab4_ex2.constraints.add(expr= model_lab4_ex2.x[2] + model_lab4_ex2.x[3] +  model_lab4_ex2.x[4] <= 100  )
model_lab4_ex2.constraints.add(expr= model_lab4_ex2.x[0] + model_lab4_ex2.x[1] >= 10  )

In [23]:
model_lab4_ex2.constraints.add(summation(data_csvfile['Cost per advt'], model_lab4_ex2.x) <= 1500000)


In [24]:
model_lab4_ex2.constraints.add(summation(data_csvfile['Num Customers'], model_lab4_ex2.x) >= 900000)


In [25]:
lower_bound = 0
upper_bound = np.inf

In [26]:
for j in col_indices:
  model_lab4_ex2.x[j].setlb(lower_bound)
  model_lab4_ex2.x[j].setub(upper_bound)

In [27]:
model_lab4_ex2.pprint()

2 Set Declarations
    constraints_index : Size=1, Index=None, Ordered=Insertion
        Key  : Dimen : Domain : Size : Members
        None :     1 :    Any :    5 : {1, 2, 3, 4, 5}
    x_index : Size=1, Index=None, Ordered=False
        Key  : Dimen : Domain : Size : Members
        None :     1 :    Any :    5 : {0, 1, 2, 3, 4}

1 Var Declarations
    x : Size=5, Index=x_index
        Key : Lower : Value : Upper : Fixed : Stale : Domain
          0 :     0 :  None :   inf : False :  True :  Reals
          1 :     0 :  None :   inf : False :  True :  Reals
          2 :     0 :  None :   inf : False :  True :  Reals
          3 :     0 :  None :   inf : False :  True :  Reals
          4 :     0 :  None :   inf : False :  True :  Reals

1 Objective Declarations
    objective : Size=1, Index=None, Active=True
        Key  : Active : Sense    : Expression
        None :   True : maximize : 35*x[0] + 57*x[1] + 48*x[2] + 20*x[3] + 15*x[4]

1 Constraint Declarations
    constraints : Siz

In [28]:
!apt-get install -y -qq coinor-cbc

Selecting previously unselected package coinor-libcoinutils3v5.
(Reading database ... 148492 files and directories currently installed.)
Preparing to unpack .../0-coinor-libcoinutils3v5_2.10.14+repack1-1_amd64.deb ...
Unpacking coinor-libcoinutils3v5 (2.10.14+repack1-1) ...
Selecting previously unselected package coinor-libosi1v5.
Preparing to unpack .../1-coinor-libosi1v5_0.107.9+repack1-1_amd64.deb ...
Unpacking coinor-libosi1v5 (0.107.9+repack1-1) ...
Selecting previously unselected package coinor-libclp1.
Preparing to unpack .../2-coinor-libclp1_1.16.11+repack1-1_amd64.deb ...
Unpacking coinor-libclp1 (1.16.11+repack1-1) ...
Selecting previously unselected package coinor-libcgl1.
Preparing to unpack .../3-coinor-libcgl1_0.59.10+repack1-1_amd64.deb ...
Unpacking coinor-libcgl1 (0.59.10+repack1-1) ...
Selecting previously unselected package coinor-libcbc3.
Preparing to unpack .../4-coinor-libcbc3_2.9.9+repack1-1_amd64.deb ...
Unpacking coinor-libcbc3 (2.9.9+repack1-1) ...
Selecting p

In [29]:
opt_cbc=SolverFactory('cbc')

In [30]:
result = opt_cbc.solve(model_lab4_ex2)
print('Solver status:', result.solver.status)
print('Solver termination condition:',result.solver.termination_condition)

Solver status: ok
Solver termination condition: optimal


In [31]:
print('\nObjective = ', model_lab4_ex2.objective())

print('\nDecision Variables')
for i in col_indices:
  print('Num slots for medium ',data_csvfile['Medium'][i], model_lab4_ex2.x[i].value)

print('\nConstraints')
model_lab4_ex2.constraints.display()


Objective =  4296.9230776

Decision Variables
Num slots for medium  Websites 4.2307692
Num slots for medium  TV 5.7692308
Num slots for medium  DailyNewspaper 65.0
Num slots for medium  WeeklyMagazine 35.0
Num slots for medium  MonthlyMagazine 0.0

Constraints
constraints : Size=5
    Key : Lower    : Body               : Upper
      1 :     None :               75.0 :      75.0
      2 :     None :              100.0 :     100.0
      3 :     10.0 :               10.0 :      None
      4 :     None : 1500000.0019999999 : 1500000.0
      5 : 900000.0 :        17882692.32 :      None


In [32]:
model_lab4_ex2.x.domain=NonNegativeIntegers

In [33]:
result=opt_cbc.solve(model_lab4_ex2)
print('Solver status:', result.solver.status)
print('Solver termination condition:',result.solver.termination_condition)

Solver status: ok
Solver termination condition: optimal


In [34]:
print('\nObjective = ', model_lab4_ex2.objective())

print('\nDecision Variables')
for i in col_indices:
  print('Num slots for medium ',data_csvfile['Medium'][i], model_lab4_ex2.x[i].value)

print('\nConstraints')
model_lab4_ex2.constraints.display()


Objective =  4294.0

Decision Variables
Num slots for medium  Websites 7.0
Num slots for medium  TV 5.0
Num slots for medium  DailyNewspaper 63.0
Num slots for medium  WeeklyMagazine 37.0
Num slots for medium  MonthlyMagazine 0.0

Constraints
constraints : Size=5
    Key : Lower    : Body       : Upper
      1 :     None :       75.0 :      75.0
      2 :     None :      100.0 :     100.0
      3 :     10.0 :       12.0 :      None
      4 :     None :  1500000.0 : 1500000.0
      5 : 900000.0 : 17465000.0 :      None


#Que.4&5

solving using approach 1

In [35]:
model_lab4_ex2.u=Var(domain=NonNegativeReals)

In [36]:
model_lab4_ex2.u.setub(18)
model_lab4_ex2.constraints.add(expr=model_lab4_ex2.u-model_lab4_ex2.x[0]-model_lab4_ex2.x[1]+model_lab4_ex2.x[2]+model_lab4_ex2.x[3]+model_lab4_ex2.x[4]>=0)
model_lab4_ex2.constraints.add(expr=model_lab4_ex2.u+ model_lab4_ex2.x[0]+model_lab4_ex2.x[1] - model_lab4_ex2.x[2] - model_lab4_ex2.x[3] - model_lab4_ex2.x[4] >=0)

In [37]:
model_lab4_ex2.pprint()

2 Set Declarations
    constraints_index : Size=1, Index=None, Ordered=Insertion
        Key  : Dimen : Domain : Size : Members
        None :     1 :    Any :    7 : {1, 2, 3, 4, 5, 6, 7}
    x_index : Size=1, Index=None, Ordered=False
        Key  : Dimen : Domain : Size : Members
        None :     1 :    Any :    5 : {0, 1, 2, 3, 4}

2 Var Declarations
    u : Size=1, Index=None
        Key  : Lower : Value : Upper : Fixed : Stale : Domain
        None :     0 :  None :    18 : False :  True : NonNegativeReals
    x : Size=5, Index=x_index
        Key : Lower : Value : Upper : Fixed : Stale : Domain
          0 :     0 :   7.0 :   inf : False : False : NonNegativeIntegers
          1 :     0 :   5.0 :   inf : False : False : NonNegativeIntegers
          2 :     0 :  63.0 :   inf : False : False : NonNegativeIntegers
          3 :     0 :  37.0 :   inf : False : False : NonNegativeIntegers
          4 :     0 :   0.0 :   inf : False : False : NonNegativeIntegers

1 Objective Declar

In [38]:
result1=opt_cbc.solve(model_lab4_ex2)
print(result1)
print(f"The solver status is: {result1.solver.status}")
print(f"The termination condition is : {result1.solver.termination_condition}")


Problem: 
- Name: unknown
  Lower bound: 3378.0
  Upper bound: 3378.0
  Number of objectives: 1
  Number of constraints: 7
  Number of variables: 5
  Number of binary variables: 0
  Number of integer variables: 5
  Number of nonzeros: 5
  Sense: maximize
Solver: 
- Status: ok
  User time: -1.0
  System time: 0.01
  Wallclock time: 0.01
  Termination condition: optimal
  Termination message: Model was solved to optimality (subject to tolerances), and an optimal solution is available.
  Statistics: 
    Branch and bound: 
      Number of bounded subproblems: 0
      Number of created subproblems: 0
    Black box: 
      Number of iterations: 1
  Error rc: 0
  Time: 0.035628318786621094
Solution: 
- number of solutions: 0
  number of solutions displayed: 0

The solver status is: ok
The termination condition is : optimal


In [39]:
print('\nObjective = ', model_lab4_ex2.objective())

print('\n Original Decision Variables')
for i in col_indices:
  print('Num slots for medium ',data_csvfile['Medium'][i], model_lab4_ex2.x[i].value)
print('Value of u = ',model_lab4_ex2.u.value)
print('\nConstraints')
model_lab4_ex2.constraints.display()


Objective =  3378.0

 Original Decision Variables
Num slots for medium  Websites 34.0
Num slots for medium  TV 0.0
Num slots for medium  DailyNewspaper 41.0
Num slots for medium  WeeklyMagazine 11.0
Num slots for medium  MonthlyMagazine 0.0
Value of u =  18.0

Constraints
constraints : Size=7
    Key : Lower    : Body       : Upper
      1 :     None :       75.0 :      75.0
      2 :     None :       52.0 :     100.0
      3 :     10.0 :       34.0 :      None
      4 :     None :  1485000.0 : 1500000.0
      5 : 900000.0 : 12095000.0 :      None
      6 :      0.0 :       36.0 :      None
      7 :      0.0 :        0.0 :      None


The Solver status is Ok and the termination condition is Optimal.
and yes we are getting optimal solution and the value of the objective funtion is:3378

**Values of Decision Variables:**

Num slots for medium  Websites =34.0

Num slots for medium  TV= 0.0

Num slots for medium  DailyNewspaper =41

Num slots for medium  WeeklyMagazine= 11

Num slots for medium  MonthlyMagazine =0.0

Value of u =  18.0


**Constraint Activity:**

constraint 1 and 7 are active as they hold at equality and rest are not active as they don't hold at equality.


Solving using approach 2

In [40]:
model_lab4_ex2.a=Var(domain=NonNegativeReals)
model_lab4_ex2.b=Var(domain=NonNegativeReals)

In [41]:
model_lab4_ex2.constraints[6].deactivate()
model_lab4_ex2.constraints[7].deactivate()

In [42]:
model_lab4_ex2.constraints.add(expr =model_lab4_ex2.x[0]+model_lab4_ex2.x[1] - model_lab4_ex2.x[2] - model_lab4_ex2.x[3] - model_lab4_ex2.x[4] - model_lab4_ex2.a + model_lab4_ex2.b ==0)
model_lab4_ex2.constraints.add(expr =model_lab4_ex2.a + model_lab4_ex2.b <=18)

In [43]:
model_lab4_ex2.pprint()

2 Set Declarations
    constraints_index : Size=1, Index=None, Ordered=Insertion
        Key  : Dimen : Domain : Size : Members
        None :     1 :    Any :    9 : {1, 2, 3, 4, 5, 6, 7, 8, 9}
    x_index : Size=1, Index=None, Ordered=False
        Key  : Dimen : Domain : Size : Members
        None :     1 :    Any :    5 : {0, 1, 2, 3, 4}

4 Var Declarations
    a : Size=1, Index=None
        Key  : Lower : Value : Upper : Fixed : Stale : Domain
        None :     0 :  None :  None : False :  True : NonNegativeReals
    b : Size=1, Index=None
        Key  : Lower : Value : Upper : Fixed : Stale : Domain
        None :     0 :  None :  None : False :  True : NonNegativeReals
    u : Size=1, Index=None
        Key  : Lower : Value : Upper : Fixed : Stale : Domain
        None :     0 :  18.0 :    18 : False : False : NonNegativeReals
    x : Size=5, Index=x_index
        Key : Lower : Value : Upper : Fixed : Stale : Domain
          0 :     0 :  34.0 :   inf : False : False : NonNega

In [44]:
result2=opt_cbc.solve(model_lab4_ex2)
print(result2)
print(f"The solver status is: {result2.solver.status}")
print(f"The termination condition is:{result2.solver.termination_condition}")


Problem: 
- Name: unknown
  Lower bound: 3378.0
  Upper bound: 3378.0
  Number of objectives: 1
  Number of constraints: 7
  Number of variables: 7
  Number of binary variables: 0
  Number of integer variables: 5
  Number of nonzeros: 5
  Sense: maximize
Solver: 
- Status: ok
  User time: -1.0
  System time: 0.01
  Wallclock time: 0.01
  Termination condition: optimal
  Termination message: Model was solved to optimality (subject to tolerances), and an optimal solution is available.
  Statistics: 
    Branch and bound: 
      Number of bounded subproblems: 0
      Number of created subproblems: 0
    Black box: 
      Number of iterations: 1
  Error rc: 0
  Time: 0.03857016563415527
Solution: 
- number of solutions: 0
  number of solutions displayed: 0

The solver status is: ok
The termination condition is:optimal


In [45]:
print('\nObjective = ', model_lab4_ex2.objective())

print('\n Original Decision Variables')
for i in col_indices:
  print('Num slots for medium ',data_csvfile['Medium'][i], model_lab4_ex2.x[i].value)
print('Value of a = ',model_lab4_ex2.a.value)
print('Value of b = ',model_lab4_ex2.b.value)
print('\nConstraints')
model_lab4_ex2.constraints.display()


Objective =  3378.0

 Original Decision Variables
Num slots for medium  Websites 34.0
Num slots for medium  TV 0.0
Num slots for medium  DailyNewspaper 41.0
Num slots for medium  WeeklyMagazine 11.0
Num slots for medium  MonthlyMagazine 0.0
Value of a =  0.0
Value of b =  18.0

Constraints
constraints : Size=9
    Key : Lower    : Body       : Upper
      1 :     None :       75.0 :      75.0
      2 :     None :       52.0 :     100.0
      3 :     10.0 :       34.0 :      None
      4 :     None :  1485000.0 : 1500000.0
      5 : 900000.0 : 12095000.0 :      None
      8 :      0.0 :        0.0 :       0.0
      9 :     None :       18.0 :      18.0


The Solver status is ok and the termination condition is Optimal and yes we are getting optimal solution and the objective function value is:3378

**Values of Decision Variables:**

Num slots for medium  Websites =34.0

Num slots for medium  TV =0.0

Num slots for medium  DailyNewspaper =41

Num slots for medium  WeeklyMagazine =11

Num slots for medium  MonthlyMagazine =0.0

Value of a =  0.0

Value of b =  18.0

**Constraint Activity:**

Constraint 1 ,8 and 9 are active as they hold at equality,rest are not active as they don't hold at equality.

#Que.6

Both approaches gave the same optimal value and same value of decision variables.

#Que.7

given that:$\ \ z_3=x_4+x_5$

we have to include the constraint $\ \ |z_3-x_3|\geq4$

this constraint can be written as $\ |x_4+x_5-x_3|\geq4$

After adding this constraint our model will be:
\begin{align}
\max \ 35 x_1 + 57 x_2 + 48 x_3 + 20 x_4 + 15 x_5, \\
{\rm{s.t.}} \; x_1 + x_2 + x_3 & \leq 75, \\
x_3 + x_4 + x_5 & \leq 100, \\
x_1 + x_2 & \geq 10, \\ 
30000 x_1 + 95000 x_2 + 10000 x_3 + 5000 x_4 + 4000 x_5 & \leq 1500000, \\
100000 x_1 + 500000 x_2 + 200000 x_3 + 45000 x_4 + 25000 x_5 & \geq 900000, \\
|x_4+x_5-x_3|\geq4,\\
x_i & \geq 0, \quad i \in \{1,2,3,4,5\}. \\
\end{align}



This formulation is not a linear program because the constraints of this model consist a modulus function which is a piecewise function. hence above program is not linear.

There are two approaches to convert the above program to a linear program:

**Approach 1:**

let $u=|x_4+x_5-x_3|$ and by the modulus function property it is clear that $u\geq0$ and $u\geq4$ and it satisfies the following inequalities:

$$u \geq x_4+x_5-x_3 \\
u \geq -x_4-x_5+x_3 $$ 
After adding these conditions our model will be:

\begin{align}
\max \ 35 x_1 + 57 x_2 + 48 x_3 + 20 x_4 + 15 x_5, \\
{\rm{s.t.}} \; x_1 + x_2 + x_3 & \leq 75, \\
x_3 + x_4 + x_5 & \leq 100, \\
x_1 + x_2 & \geq 10, \\ 
30000 x_1 + 95000 x_2 + 10000 x_3 + 5000 x_4 + 4000 x_5 & \leq 1500000, \\
100000 x_1 + 500000 x_2 + 200000 x_3 + 45000 x_4 + 25000 x_5 & \geq 900000, \\x_4+x_5-x_3\leq u,\\ -x_4-x_5+x_3\leq u,\\u\geq 4,\\
x_i & \geq 0, \quad i \in \{1,2,3,4,5\}. \\
\end{align}



**Approach2:-**

let $x_4+x_5-x_3=a-b$

then $|x_4+x_5-x_3|=|a-b|=a+b\ $ where $a\geq0$ and $b\geq0$

now our model will be:
\begin{align}
\max \ 35 x_1 + 57 x_2 + 48 x_3 + 20 x_4 + 15 x_5, \\
{\rm{s.t.}} \; x_1 + x_2 + x_3 & \leq 75, \\
x_3 + x_4 + x_5 & \leq 100, \\
x_1 + x_2 & \geq 10, \\ 
30000 x_1 + 95000 x_2 + 10000 x_3 + 5000 x_4 + 4000 x_5 & \leq 1500000, \\
100000 x_1 + 500000 x_2 + 200000 x_3 + 45000 x_4 + 25000 x_5 & \geq 900000, \\
x_4+x_5-x_3-a+b=0\\
a+b\geq 4,\\
x_i & \geq 0, \quad i \in \{1,2,3,4,5\}. \\ a,b \geq 0
\end{align}